KsApi/GraphSchema.swift

```diff
@@ -0,0 +1,398 @@
+import Prelude
+
+/// Graph Response
+
+struct GraphResponse<T: Decodable>: Decodable {
+  let data: T?
+}
+
+/// Base Query Types
+
+extension Never: CustomStringConvertible {
+  public var description: String {
+    fatalError()
+  }
+}
+
+public protocol QueryType: CustomStringConvertible, Hashable {}
+
+extension QueryType {
+  public var hashValue: Int {
+    return description.hashValue
+  }
+}
+
+public enum Connection<Q: QueryType> {
+  case pageInfo(NonEmptySet<PageInfo>)
+  case edges(NonEmptySet<Edges<Q>>)
+  case nodes(NonEmptySet<Q>)
+  case totalCount
+}
+
+public func == <Q: QueryType>(lhs: Q, rhs: Q) -> Bool {
+  return lhs.hashValue == rhs.hashValue
+}
+
+public func join<S: Sequence>(_ args: S, _ separator: String = " ") -> String
+  where S.Iterator.Element: QueryType {
+    return args.map { $0.description }.sorted().joined(separator: separator)
+}
+
+public func join<Q: QueryType>(_ nodes: NonEmptySet<Q>, _ separator: String = " ") -> String {
+  return join(Array(nodes))
+}
+
+public func decodeBase64(_ input: String) -> String? {
+  return Data(base64Encoded: input)
+    .flatMap { String(data: $0, encoding: .utf8) }
+}
+
+public func decompose(id: String) -> (String, Int)? {
+
+  return decodeBase64(id)
+    .flatMap { id -> (String, Int)? in
+      let pair = id.split(separator: "-", maxSplits: 1)
+      return zip(pair.first.map(String.init), pair.last.flatMap { Int($0) })
+  }
+}
+
+public struct RelayId: Swift.Decodable {
+  let id: String
+}
+
+extension RelayId: ExpressibleByStringLiteral {
+  public init(unicodeScalarLiteral value: String) {
+    self.init(id: value)
+  }
+  public init(extendedGraphemeClusterLiteral value: String) {
+    self.init(id: value)
+  }
+  public init(stringLiteral value: String) {
+    self.init(id: value)
+  }
+}
+
+public enum QueryArg<T: CustomStringConvertible> {
+  case after(String)
+  case before(String)
+  case first(Int)
+  case last(Int)
+  case arg(T)
+}
+
+public enum EdgesContainerBody<Q: QueryType> {
+  case pageInfo(NonEmptySet<PageInfo>)
+  case edges(NonEmptySet<Edges<Q>>)
+}
+
+public enum PageInfo: String {
+  case endCursor
+  case hasNextPage
+  case hasPreviousPage
+  case startCursor
+}
+
+public enum Edges<Q: QueryType> {
+  case cursor
+  case node(NonEmptySet<Q>)
+}
+
+public enum Nodes<Q: QueryType> {
+  case nodes(NonEmptySet<Q>)
+}
+
+//public struct GraphError: Error {
```

Comment:
Still needed?

Reply:
The GraphError struct is no longer needed, so I will delete it